### PR TITLE
Rustup to rustc 1.28.0-nightly (952f344cd 2018-05-18)

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

**Comment by oli-obk** [comment]
run-pass/closures is failing:

> librustc_mir/interpret/eval_context.rs:1111: value_to_primval can't work with fat pointers

**Comment by bjorn3** [comment]
Backtrace:

```
[...]
  13: rustc::session::bug_fmt
  14: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::value_to_primval
  15: rustc_mir::interpret::operator::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::intrinsic_with_overflow
  16: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_rvalue_into_place
  17: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::step
  18: miri::eval_main
  19: <miri::MiriCompilerCalls as rustc_driver::CompilerCalls<'a>>::build_controller::{{closure}}
[...]
```

Doesn't seem to be related to the changes in this PR

**Comment by oli-obk** [comment]
Indeed. I'll see what I can do on the rustc side

**Comment by oli-obk** [comment]
Maybe related : https://github.com/rust-lang/rust/issues/50706

File: src/bin/miri.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ impl<'a> CompilerCalls<'a> for MiriCompilerCalls {
     }
     fn late_callback(
         &mut self,
-        trans: &TransCrate,
+        trans: &CodegenBackend,
```
Comment (eddyb) [inline src/bin/miri.rs]: This argument should be renamed `codegen_backend`.

**Comment by oli-obk** [comment]
There's a bunch of new failures: `Size::from_bytes: 18446744073709551614 bytes in bits doesn't fit in u64`

**Comment by bjorn3** [comment]
Those failures are inside rustc:

```
   7: rustc_target::abi::Size::from_bytes
   8: rustc::mir::interpret::value::Pointer::wrapping_signed_offset
   9: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, miri::Evaluator<'tcx>> as miri::helpers::EvalContextExt<'tcx>>::wrapping_pointer_offset
  10: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, miri::Evaluator<'tcx>> as miri::intrinsic::EvalContextExt<'tcx>>::call_intrinsic
  11: rustc_mir::interpret::terminator::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_fn_call
  12: rustc_mir::interpret::terminator::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_terminator
  13: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::step
  14: miri::eval_main
```

**Comment by bjorn3** [comment]
screenshot_2018-05-20 miri tests run-pass ptr_arith_offset_overflow rs

**Comment by oli-obk** [comment]
We need to allow pointers one past the end of an allocation:

> pointer computed at offset 10, outside bounds of allocation 462 which has size 9

**Comment by RalfJung** [comment]
Nope, if the size is 9 then 10 is *two* past the end.

**Comment by bjorn3** [comment]
rust-lang/rust#50967 required a huge amount of renaming and I still haven't got the majority of the tests passing after several hours :(

**Comment by bjorn3** [comment]
I get many error like:

```
error: attempted to read undefined bytes

ERROR 2018-05-26T14:44:50Z: rustc_mir::interpret::eval_context: 

An error occurred in miri:
backtrace frames: 27
0: backtrace::backtrace::trace::h85503c2e4e338460
	at <unknown_file>
1: backtrace::capture::Backtrace::new_unresolved::h51246d60a53a8bdc
	at <unknown_file>
2: _$LT$rustc..mir..interpret..error..EvalError$LT$$u27$tcx$GT$$u20$as$u20$core..convert..From$LT$rustc..mir..interpret..error..EvalErrorKind$LT$$u27$tcx$C$$u20$u64$GT$$GT$$GT$::from::h4d30d3209764fd49
	at <unknown_file>
3: rustc::mir::interpret::value::Scalar::to_bits::h85909ebea5cc3804
	at <unknown_file>
4: rustc_mir::interpret::terminator::_$LT$impl$u20$rustc_mir..interpret..eval_context..EvalContext$LT$$u27$a$C$$u20$$u27$mir$C$$u20$$u27$tcx$C$$u20$M$GT$$GT$::eval_terminator::h69abb3bb79aab288
	at /Users/travis/build/rust-lang/rust/src/librustc_mir/interpret/terminator/mod.rs:47
5: rustc_mir::interpret::step::_$LT$impl$u20$rustc_mir..interpret..eval_context..EvalContext$LT$$u27$a$C$$u20$$u27$mir$C$$u20$$u27$tcx$C$$u20$M$GT$$GT$::terminator::hdd1ce4d831f07250
	at /Users/travis/build/rust-lang/rust/src/librustc_mir/interpret/step.rs:109
6: rustc_mir::interpret::step::_$LT$impl$u20$rustc_mir..interpret..eval_context..EvalContext$LT$$u27$a$C$$u20$$u27$mir$C$$u20$$u27$tcx$C$$u20$M$GT$$GT$::step::h7b4b7c429541cd9b
	at /Users/travis/build/rust-lang/rust/src/librustc_mir/interpret/step.rs:43
7: miri::eval_main::run_main::h9c72be881571a562
	at src/lib.rs:232
8: miri::eval_main::h9d8ab3b0d25eb037
	at src/lib.rs:245
9: miri::after_analysis::h1dc1a8379cb87900
	at src/bin/miri.rs:147
10: _$LT$miri..MiriCompilerCalls$u20$as$u20$rustc_driver..CompilerCalls$LT$$u27$a$GT$$GT$::build_controller::_$u7b$$u7b$closure$u7d$$u7d$::hbed7da9498dc169f
	at src/bin/miri.rs:88
11: rustc::ty::context::tls::with_context::h434a73bf87917666
	at <unknown_file>
12: rustc_driver::driver::compile_input::_$u7b$$u7b$closure$u7d$$u7d$::h54692425a596d732
	at <unknown_file>
13: rustc::ty::context::tls::enter_context::hdfa81ea4757b9ad4
	at <unknown_file>
14: _$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::with::hfbf5029193cab3f2
	at <unknown_file>
15: rustc::ty::context::TyCtxt::create_and_enter::hfc13991376fb466d
	at <unknown_file>
16: rustc_driver::driver::compile_input::he7ba295ba0b6472f
	at <unknown_file>
17: rustc_driver::run_compiler_with_pool::h39dace11c41da12a
	at <unknown_file>
18: syntax::with_globals::h39d73970aba4a7ab
	at <unknown_file>
19: rustc_driver::run_compiler::hf76541365053c7e0
	at <unknown_file>
20: miri::main::hd88a1c4b352e05a3
	at src/bin/miri.rs:237
21: std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::hf5b1442a7c729874
	at /Users/travis/build/rust-lang/rust/src/libstd/rt.rs:74
22: std::panicking::try::do_call::hfd75cf9e05cc5943 (.llvm.9696875478146817961)
	at <unknown_file>
23: __rust_maybe_catch_panic
	at <unknown_file>
24: std::rt::lang_start_internal::h36f7afe722d0078f
	at <unknown_file>
25: std::rt::lang_start::he4af6349eec3f416
	at /Users/travis/build/rust-lang/rust/src/libstd/rt.rs:74
26: main
	at <unknown_file>

error[E0080]: constant evaluation error
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/result.rs:747:13
    |
747 |             Ok(t) => t,
    |             ^^^^^ attempted to read undefined bytes
    |
note: inside call to `<std::result::Result<T, E>><(), std::collections::CollectionAllocErr>::unwrap_or_else::<[closure@DefId(4/1:226 ~ alloc[e00a]::raw_vec[0]::{{impl}}[0]::allocate_in[0]::{{closure}}[1])]>`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/raw_vec.rs:92:13
    |
92  |             alloc_guard(alloc_size).unwrap_or_else(|_| capacity_overflow());
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<alloc::raw_vec::RawVec<T, A>><(), std::heap::Global>::allocate_in`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/raw_vec.rs:146:9
    |
146 |         RawVec::allocate_in(cap, false, Global)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<alloc::raw_vec::RawVec<T>><()>::with_capacity`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/vec.rs:362:18
    |
362 |             buf: RawVec::with_capacity(capacity),
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<std::vec::Vec<T>><()>::with_capacity`
   --> /Users/bjorn/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/binary_heap.rs:342:28
    |
342 |         BinaryHeap { data: Vec::with_capacity(capacity) }
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<std::collections::BinaryHeap<T>><()>::with_capacity`
   --> tests/run-pass/zero-sized-binary-heap-push.rs:18:26
    |
18  |         let mut tester = BinaryHeap::with_capacity(len);
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by bjorn3** [comment]
Closing in favour of #379

**Pull request closed without merging**